### PR TITLE
Developer documentation update involving module invocation

### LINKED ISSUE #55418: Changes to the unsafe marking
##### SUMMARY
If I'm reading https://github.com/ansible/ansible/pull/55398 correctly, it solves the auto-interpreter not marking the returned information as Unsafe and also marks all facts as unsafe outside of the action plugins.  If that's true, then there's two changes that seem like they should be made:

* The changelog needs to specify that this fixes an insecurity in autointerpreter that was only present in 2.8.0alpha1 and 2.8.0beta1.
* The wrap_var call in lib/ansible/plugins/action/__init__.py is no longer needed an should be removed.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
* lib/ansible/plugins/action/__init__.py
* changelogs/fragments/ensure_facts_safe.yml

##### ANSIBLE VERSION

```paste below
2.8
```

##### CONFIGURATION

```paste below
N/A
```

##### OS / ENVIRONMENT

N/A

##### STEPS TO REPRODUCE

```yaml
N/A
```

##### EXPECTED RESULTS

N/A

##### ACTUAL RESULTS

```paste below
N/A
```

@bcoca

### PR DESCRIPTION
**Pull request by abadger**
* Update for the 2.7 change to AnsiballZ which invokes modules with one
  less Python interpreter
* Add a section on how module results are returned.
* Add a section on trust between modules, action plugins, and the
  executor.

Fixes #55418

##### ISSUE TYPE

- Docs Pull Request

##### COMPONENT NAME

docs/docsite/rst/dev_guide/developing_program_flow_modules.rst

##### ADDITIONAL INFORMATION

Adding information on why we mark strings in results as unsafe at both the action plugin and executor level was the primary motivator for this update.  Once I was looking at the file, though, I discovered that the document hadn't been updated for the 2.7 changes to AnsiballZ either so I added those.

I'm not sure if the information on marking strings belongs in this doc.  If you can think of a better place I'm open to moving it.

/cc @acozine @nitzmahone

### CODE REVIEW

File: docs/docsite/rst/dev_guide/developing_program_flow_modules.rst [hunk 1]
```diff
@@ -326,19 +326,25 @@ in :file:`ansible/module_utils` that are imported by the module, and some
...
 and places the zipfile into a temp directory on the managed node.  It then
-extracts just the ansible module script from the zip file and places that in
-the temporary directory as well.  Then it sets the PYTHONPATH to find python
-modules inside of the zip file and invokes :command:`python` on the extracted
-ansible module.
+extracts just the Ansible module script from the zip file and places that in
+the temporary directory as well.  Then it sets the PYTHONPATH to find Python
+modules inside of the zip file and imports the Ansible module as the special name, ``__main__``.
+Importing it as ``__main__`` causes Python to think that it is executing a script rather than simply
+importing a module, thus running the code within.
```
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]:
```suggestion
importing a module. This lets Ansible run both the wrapper script and the module code in a single copy of Python on the remote machine.
```

File: docs/docsite/rst/dev_guide/developing_program_flow_modules.rst [hunk 2]
```diff
@@ -5,42 +5,37 @@
 Ansible module architecture
 ***************************
 
-This in-depth dive helps you understand Ansible's program flow to execute
-modules. It is written for people working on the portions of the Core Ansible
-Engine that execute a module. Those writing Ansible Modules may also find this
-in-depth dive to be of interest, but individuals simply using Ansible Modules
-will not likely find this to be helpful.
+If you're working on Ansible's Core code, or writing an Ansible module, this deep dive helps you understand how Ansible's program flow executes
```
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: i would add 'developing action plugins' to the reasons you would want to read this

File: docs/docsite/rst/dev_guide/developing_program_flow_modules.rst [hunk 3]
```diff
@@ -49,23 +44,20 @@ into its final location, sets file permissions, and so on.
 New-style modules
 -----------------
 
-All of the modules that ship with Ansible fall into this category.
+All of the modules that ship with Ansible fall into this category. All official modules (shipped with Ansible) use either Python or PowerShell.
```
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: I would prefix with: 'While you can write modules in any language, "

File: docs/docsite/rst/dev_guide/developing_program_flow_modules.rst [hunk 4]
```diff
@@ -5,42 +5,37 @@
...
-always execute on the controller and are sometimes able to do all work there
-(for instance, the ``debug`` Action Plugin which prints some text for the user to
-see or the ``assert`` Action Plugin which can test whether several values in
-a playbook satisfy certain criteria.)
-
-More often, Action Plugins set up some values on the controller, then invoke an
-actual module on the managed node that does something with these values.  An
-easy to understand version of this is the :ref:`template Action Plugin
-<template_module>`.  The :ref:`template Action Plugin <template_module>` takes values from
+Action plugins look like modules to anyone writing a playbook. They are documented as modules. However, they behave differently on the back end. Action plugins always execute on the controller.
```
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: 'they are documented as modules' can be read in 2 ways, 'their usage documentation lives inside a module' or 'the documentation makes no distinction on module vs action plugin', the latter being not completely true since we do reflect that 'this module has an action plugin' and that they are diff types than modules. 

I would rephrase:
"The 'action' of a task is normally assumed to be the same as a module of the same name, but there is always an 'action plugin' involved, sometimes they just execute the  module, other times they do all the work and the module is just there as documentation, in other cases it is something in between."

"Then the paragraph below is good with examples, i would also add "then modules that don't have a corresponding action plugin are executed by the `<link normal section below>`'normal' action plugin`</link>`"

File: docs/docsite/rst/dev_guide/developing_program_flow_modules.rst [hunk 5]
```diff
@@ -185,83 +175,85 @@ Executor/task_executor
...
+The ``normal`` action plugin
+----------------------------
 
-The ``normal`` action plugin executes the module on the remote host.  It is
+The ``normal`` action plugin executes the module on the remote host. It is
 the primary coordinator of much of the work to actually execute the module on
 the managed machine.
 
-* It takes care of creating a connection to the managed machine by
+* It creates a connection to the managed machine by
```
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: It uses a connection plugin to create a connection, mostly indirectly by asking the connection plugin to transfer the module or other files ... it does not really create it itself directly.

File: docs/docsite/rst/dev_guide/developing_program_flow_modules.rst [hunk 6]
```diff
@@ -351,128 +345,120 @@ the zipfile as well.
...
+changes to be made to templated files. To set, pass the ``--diff`` command line
+option. To access in a module, instantiate an `AnsibleModule` and access
 :attr:`AnsibleModule._diff`.
 
 _ansible_verbosity
 ^^^^^^^^^^^^^^^^^^
 
-This value could be used for finer grained control over logging. However, it
-is currently unused.
+Unused. This value could be used for finer grained control over logging.
```
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_program_flow_modules.rst:421]: missing several _ansible_module_name, _ansible_string_conversion_action, _ansible_keep_remote_files, _ansible_socket, _ansible_shell_executable, _ansible_tmpdir, _ansible_remote_tmp

**Review by bcoca — CHANGES_REQUESTED** [review]

File: docs/docsite/rst/dev_guide/developing_program_flow_modules.rst (hunk 6 above)
Comment (acozine) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst:421]: I'll make this list comprehensive in a separate PR.  I've opened https://github.com/ansible/ansible/issues/55935 so I can keep track of that task.

**Review by bcoca — APPROVED** [review]
